src/librbd/cache/pwl/Types.h

```diff
@@ -212,27 +212,32 @@ struct WriteLogCacheEntry {
   uint64_t write_sequence_number = 0;
   uint64_t image_offset_bytes;
   uint64_t write_bytes;
-  #ifdef WITH_RBD_RWL
-  TOID(uint8_t) write_data;
-  #endif
-  #ifdef WITH_RBD_SSD_CACHE
-  uint64_t write_data_pos = 0; /* SSD data offset */
-  #endif
   union {
-    uint8_t flags;
+    uint64_t flags;
     struct {
-      uint8_t entry_valid :1; /* if 0, this entry is free */
-      uint8_t sync_point :1;  /* No data. No write sequence number. Marks sync
+      uint64_t entry_valid :1; /* if 0, this entry is free */
+      uint64_t sync_point :1;  /* No data. No write sequence number. Marks sync
                                  point for this sync gen number */
-      uint8_t sequenced :1;   /* write sequence number is valid */
-      uint8_t has_data :1;    /* write_data field is valid (else ignore) */
-      uint8_t discard :1;     /* has_data will be 0 if this is a discard */
-      uint8_t writesame :1;   /* ws_datalen indicates length of data at write_bytes */
+      uint64_t sequenced :1;   /* write sequence number is valid */
+      uint64_t has_data :1;    /* write_data field is valid (else ignore) */
+      uint64_t discard :1;     /* has_data will be 0 if this is a discard */
+      uint64_t writesame :1;   /* ws_datalen indicates length of data at write_bytes */
     };
   };
-  uint32_t ws_datalen = 0;  /* Length of data buffer (writesame only) */
-  uint32_t entry_index = 0; /* For debug consistency check. Can be removed if
+  uint64_t ws_datalen = 0;  /* Length of data buffer (writesame only) */
+  uint64_t entry_index = 0; /* For debug consistency check. Can be removed if
                              * we need the space */
+  union {
+    /* RWL */
+    struct {
+      PMEMoid write_data;
+    };
```

Comment:
@CongMinYin Does this work if code is compiled only with `WITH_RBD_SSD_CACHE`?

Reply:
> @CongMinYin Does this work if code is compiled only with `WITH_RBD_SSD_CACHE`?

If change to union, it is independent of compilation options.

Reply:
I just tried to compile this PR. This PR does not compile in `WITH_RBD_SSD_CACHE` mode. That was the reason for using the `ifdef` condition initially.

```
../src/librbd/cache/pwl/Types.h:233:7: error: 'PMEMoid' does not name a type
```

Reply:
> ```
> ../src/librbd/cache/pwl/Types.h:233:7: error: 'PMEMoid' does not name a type
> ```
 Get your point, now I understand why use `#ifdef`. The available solution is to keep `#ifdef`, but leave it in the `union`, but remove `#ifdef` about `ssd`, because ssd use encode/decode, It will not be affected. So that, there are 2 case:
1) for rwl side, including `WITH_RBD_RWL`, union length is the bigger one. Whether the `WITH_RBD_SSD_CACHE` is compiled or not, union length is fixed. 
2) for ssd side, including `WITH_RBD_SSD_CACHE`,  `ssd` use encode/decode,  doesn't matter.

Reply:
I don't like this ` #ifdef WITH_RBD_RWL` in on-disk structure. I understand the reason and it seems not to cause the issue, as you explained. Still it looks at least very confusing.

And I am not sure it is a good idea to use "external" structure in on-disk structure. Are you sure `PMEMoid` will not change in the future? How are you going to track this?

Using encode/decode seems like the right solution indeed. Though I will not insist on changing this in this PR. I will not block the PR. Just wanted to bring this for discussion.

Reply:
> I don't like this ` #ifdef WITH_RBD_RWL` in on-disk structure. I understand the reason and it seems not to cause the issue, as you explained. Still it looks at least very confusing.
> 
> And I am not sure it is a good idea to use "external" structure in on-disk structure. Are you sure `PMEMoid` will not change in the future? How are you going to track this?
> 
> Using encode/decode seems like the right solution indeed. Though I will not insist on changing this in this PR. I will not block the PR. Just wanted to bring this for discussion.

I agree with you. This commit is quite old and not rigorous. As you pointed out, the external data type is unreliable and may be changed. Another side, we planned to change the libpmemobj library to libpmem to remove external types, use addresses and offsets, and do space management by ourselves, https://github.com/ceph/ceph/pull/41432.
@ideepika  So I want to remove the change of hard disk format,  keep version and memory format change. Hard disk format will be better solved in the future.